timemachine/ff/handlers/nonbonded.py

```diff
@@ -506,6 +506,96 @@ class AM1BCCSolventHandler(AM1BCCHandler):
     pass
 
 
+from openmm.app.forcefield import ForceField
+
+
+class ExtendedForceField(ForceField):
+    pass
+
+
+import ast
+
+
+class EnvironmentBCCHandler(SerializableMixIn):
+    """
+    Applies BCCs to residues in a forcefield.
+    """
+
+    def __init__(self, patterns, params, protein_ff_name, water_ff_name):
+        self.patterns = patterns
+        # self.params = np.array(params)
+        self.params = None
+        self.env_ff = ForceField(protein_ff_name, water_ff_name)
+
+        self.res_to_bonds_to_param_idxs = dict()
+        for param_idx, pattern in enumerate(self.patterns):
+            res = pattern.split()
+            res_name = res[0]
+            bonds = ast.literal_eval(" ".join(res[1:]))
+            if res_name not in self.res_to_bonds_to_param_idxs:
+                self.res_to_bonds_to_param_idxs[res_name] = dict()
+            for bond in bonds:
+                self.res_to_bonds_to_param_idxs[res_name][bond] = param_idx
+
+        # print(self.res_to_bonds)
+        # assert 0
+
+    def parameterize(self, params, topology):
+        # params here refers to *bcc* params, not per atom charges
+        # computes offsets that we can apply to each atom in topology.
+
+        residueTemplates = dict()
+        ignoreExternalBonds = False
+        data = ForceField._SystemData(topology)
+        templateForResidue = self.env_ff._matchAllResiduesToTemplates(
+            data, topology, residueTemplates, ignoreExternalBonds
+        )
+
+        total = 0
+        bond_idxs = []
+
+        # generate bcc deltas
+        bond_deltas = []
+        for index, val in enumerate(templateForResidue):
+            # get bonds within each residue
+            for src, dst in val.bonds:
+                bond_idxs.append([src + total, dst + total])
+
+                if (src, dst) in self.res_to_bonds_to_param_idxs[val.name]:
+                    delta = params[self.res_to_bonds_to_param_idxs[val.name][(src, dst)]]
+                    bond_deltas.append(delta)
+                elif (dst, src) in self.res_to_bonds_to_param_idxs[val.name]:
+                    # flip sign
+                    delta = params[self.res_to_bonds_to_param_idxs[val.name][(dst, src)]]
+                    bond_deltas.append(-delta)
+                else:
+                    assert 0
```

Comment:
nit[performance / modularity]: might want to adjust this bit to compute an index array `param_idxs` instead of directly forming a list of `bond_deltas`?

so that we can split this into two functions: one that's a possibly expensive function of the topology alone (returning `initial_charges, bond_idxs, param_idxs`, contains lots of non-differentiable stuff)

and a simpler/cheaper/differentiable function of the parameters
`bond_deltas = params[param_idxs]; return apply_bond_charge_corrections(initial_charges, bond_idxs, bond_deltas, ...)`

(since we will call this multiple times with the same `topology` and varying `params`)

Reply:
Resolved